runtime/lua/vim/lsp/buf.lua

```diff
@@ -192,17 +192,21 @@ end
 ---     - range (table|nil) Range to format.
 ---         Table must contain `start` and `end` keys with {row,col} tuples using
 ---         (1,0) indexing.
+---         Can also be a list of tables that contain `start` and `end` keys as described above,
+---         in which case textDocument/rangesFormatting support is required.
 ---         Defaults to current selection in visual mode
 ---         Defaults to `nil` in other modes, formatting the full buffer
 function M.format(options)
   options = options or {}
   local bufnr = options.bufnr or api.nvim_get_current_buf()
   local mode = api.nvim_get_mode().mode
   local range = options.range
-  if not range and mode == 'v' or mode == 'V' then
+  local ranges = (range and #range ~= 0 and type(range[1]) == 'table') and range
+  if not range and not ranges and mode == 'v' or mode == 'V' then
     range = range_from_selection(bufnr, mode)
   end
-  local method = range and ms.textDocument_rangeFormatting or ms.textDocument_formatting
+  local method = (ranges and ms.textDocument_rangesFormatting) or
+      (range and ms.textDocument_rangeFormatting or ms.textDocument_formatting)
```

Comment:
I'm not entirely sure, but if `range` is `nil` here then `method` won't be assigned. Could you please check this?

Reply:
`:lua print((nil and "a") or (nil and "b" or "c"))` prints `c`, so I think this should work

Reply:
Thanks for checking!